File: plot_covariance.m
function plot_covariance(XBS,COVARIANCE,LOCALIZATION,INCREMENT,lon,lat,LonI,LatJ,Variable)

ms=14; %Marker size
fs=[0 0 30 30];

coast=load('coast');
coast.long(coast.long < 0)=coast.long(coast.long < 0)+360;
for ii=1:length(coast.long)-1
    if( abs(coast.long(ii) - coast.long(ii+1)) > 180)
        coast.long(ii)=NaN;
        coast.long(ii+1)=NaN;
    end
end


if(Variable==1)
    VAR='U';
elseif( Variable==2)
    VAR='V';
elseif( Variable==3)
    VAR='PHI';
end

%==========================================================================
%                     U VARIANCE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position', fs)

scale=max(max(abs(XBS(:,:,1))))/5;
levels=[0 1 2 3 4 5]*scale;
colors=[2 32 34 36 38];
contourf(lon,lat,XBS(:,:,1),levels);
hold on
plot_jrcol(levels,colors);

plot(coast.long,coast.lat,'k','LineWidth',2)
title(['U Variance (m/s)'],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)

%==========================================================================
%                     V VARIANCE
%==========================================================================


hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position', fs)

scale=max(max(abs(XBS(:,:,2))))/5;
levels=[0 1 2 3 4 5]*scale;
colors=[2 32 34 36 38];
contourf(lon,lat,XBS(:,:,2),levels);
hold on
plot_jrcol(levels,colors);

plot(coast.long,coast.lat,'k','LineWidth',2)
title(['V Variance (m/s)'],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')

%==========================================================================
%                     PHI VARIANCE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position', fs)

scale=max(max(abs(XBS(:,:,3))))/5;
levels=[0 1 2 3 4 5]*scale;
colors=[2 32 34 36 38];
contourf(lon,lat,XBS(:,:,3),levels);
hold on
plot_jrcol(levels,colors);

plot(coast.long,coast.lat,'k','LineWidth',2)
title(['PHI Variance (m)'],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')


%==========================================================================
%                     VAR - U COVARIANCE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position', fs)

scale=max(max(abs(COVARIANCE(:,:,1))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,COVARIANCE(:,:,1),levels);
hold on
plot_jrcol(levels,colors);
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
plot(coast.long,coast.lat,'k','LineWidth',2)
title(['Covariance ' VAR '-U '],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')


%==========================================================================
%                     VAR - V COVARIANCE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(COVARIANCE(:,:,2))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,COVARIANCE(:,:,2),levels);
plot_jrcol(levels,colors);
hold on
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
plot(coast.long,coast.lat,'k','LineWidth',2)
%quiver(lon,lat,XBM(:,:,1,end)-XTMEAN(:,:,1),-XBM(:,:,2,end)+XTMEAN(:,:,2))
title(['Covariance ' VAR '-V '],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')


%==========================================================================
%                     VAR - PHI COVARIANCE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(COVARIANCE(:,:,3))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,COVARIANCE(:,:,3),levels);
plot_jrcol(levels,colors);
hold on
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
plot(coast.long,coast.lat,'k','LineWidth',2)
%quiver(lon,lat,XBM(:,:,1,end)-XTMEAN(:,:,1),-XBM(:,:,2,end)+XTMEAN(:,:,2))
title(['Covariance ' VAR '-PHI'],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')

%==========================================================================
%                     VAR - PHI , VAR - U , VAR - V COVARANICE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(COVARIANCE(:,:,3))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,COVARIANCE(:,:,3),levels);
plot_jrcol(levels,colors);
hold on
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
quiver(lon,lat,COVARIANCE(:,:,1),COVARIANCE(:,:,2))
plot(coast.long,coast.lat,'k','LineWidth',2)
title(['Covariance ' VAR '-PHI (shaded), ' VAR '-U ; ' VAR '-V (vectors)'],'FontSize',15)
plot(coast.long,coast.lat,'k','LineWidth',2)
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')


%==========================================================================
%                     VAR - U LOCALIZED COVARIANCE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(COVARIANCE(:,:,1))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,COVARIANCE(:,:,1).*LOCALIZATION,levels);
hold on
plot_jrcol(levels,colors);
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
plot(coast.long,coast.lat,'k','LineWidth',2)
title(['Localized covariance ' VAR '-U'],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')

%==========================================================================
%                    VAR - V LOCALIZED COVARIANCE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(COVARIANCE(:,:,2))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,COVARIANCE(:,:,2).*LOCALIZATION,levels);
plot_jrcol(levels,colors);
hold on
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
plot(coast.long,coast.lat,'k','LineWidth',2)
%quiver(lon,lat,XBM(:,:,1,end)-XTMEAN(:,:,1),-XBM(:,:,2,end)+XTMEAN(:,:,2))
title(['Localized covariance ' VAR '-V'],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')

%==========================================================================
%                     VAR - PHI LOCALIZED COVARIANCE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(COVARIANCE(:,:,3))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,COVARIANCE(:,:,3).*LOCALIZATION,levels);
plot_jrcol(levels,colors);
hold on
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
plot(coast.long,coast.lat,'k','LineWidth',2)
%quiver(lon,lat,XBM(:,:,1,end)-XTMEAN(:,:,1),-XBM(:,:,2,end)+XTMEAN(:,:,2))
title(['Localized covariance ' VAR '-PHI'],'FontSize',15);
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')

%==========================================================================
%           VAR - PHI , VAR - U , VAR - V LOCALIZED COVARANICE
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(COVARIANCE(:,:,3))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,COVARIANCE(:,:,3).*LOCALIZATION,levels);
plot_jrcol(levels,colors);
hold on
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
quiver(lon,lat,COVARIANCE(:,:,1).*LOCALIZATION,COVARIANCE(:,:,2).*LOCALIZATION)
plot(coast.long,coast.lat,'k','LineWidth',2)
title(['Localized covariance ' VAR '-PHI (shaded), ' VAR '-U ; ' VAR '-V (vectors)'],'FontSize',15)
plot(coast.long,coast.lat,'k','LineWidth',2)
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')


%==========================================================================
%   SINGLE OBSERVATION ANALYSIS INCREMENT 
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(INCREMENT(:,:,3))))/5;
levels=[-5 -4 -3 -2 -1 1 2 3 4 5]*scale;
colors=[48 46 44 42 2 32 34 36 38];
contourf(lon,lat,INCREMENT(:,:,3),levels);
plot_jrcol(levels,colors);
hold on
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
quiver(lon,lat,INCREMENT(:,:,1),INCREMENT(:,:,2))
plot(coast.long,coast.lat,'k','LineWidth',2)
title('Analysis Increment in PHI (shaded), U;V (vector)','FontSize',15)
plot(coast.long,coast.lat,'k','LineWidth',2)
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')



%==========================================================================
%   SINGLE LOCALIZATION
%==========================================================================

hFig=figure('Menubar','none'); %,'Visible','Off');
set(hFig, 'Units', 'centimeters', 'Position',fs)

scale=max(max(abs(LOCALIZATION)))/5;
levels=[0 1 2 3 4 5]*scale;
colors=[2 42 44 46 48];
contourf(lon,lat,LOCALIZATION,levels);
plot_jrcol(levels,colors);
hold on
plot(LonI,LatJ,'ok','MarkerSize',ms,'MarkerFaceColor','k')
plot(coast.long,coast.lat,'k','LineWidth',2)
title('Localization function','FontSize',15)
plot(coast.long,coast.lat,'k','LineWidth',2)
colorbar('YTick',levels);
xlabel('Lon','FontSize',15);ylabel('Lat','FontSize',15)
set(gca,'FontSize',15)
axis('square')

File: PYTHON/plot_covariance.py
import os
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.basemap import Basemap

def set_env_map(fs, lon, lat, LonI, LatJ, marker=False):

    fig, ax = plt.subplots(figsize=(14, 10))
    m = Basemap(projection='mill', lon_0=180)
    m.drawcoastlines()
    m.drawparallels(np.arange(-80, 90, 20), labels=[1, 0, 0, 0], fontsize=fs)
    m.drawmeridians(np.arange(m.lonmin, m.lonmax+30, 40), labels=[0, 0, 0, 1], fontsize=fs)
    x, y = m(lon, lat)
    Lon, Lat = m(LonI, LatJ)
    #ax.set_xlabel('Lon', labelpad=20, fontsize=fs)
    #ax.set_ylabel('Lat', labelpad=30, fontsize=fs)
    if marker:
        ax.plot(Lon, Lat, marker='.', markersize=25, color="red")
    return fig, ax, m, x, y

def plot_covariance(outdir, XMEAN, ANOM, XSPREAD, COVARIANCE, LOCALIZATION, INCREMENT, lon, lat, LonI, LatJ, Variable):

    #Create output directory
    if not os.path.exists(outdir):
        os.makedirs(outdir)

    # Palette
    seq = plt.get_cmap('YlGn')
    div = plt.get_cmap('PRGn')

    if Variable == 0:
        VAR = 'U'
        wscale = 0.05
        wkey = 0.03
    elif Variable == 1:
        VAR ='V'
        wscale = 0.05
        wkey = 0.03
    elif Variable == 2:
        VAR ='PHI'
        wscale = 0.5
        wkey = 0.3

    #Graphic parameters
    fstitle = 22
    fs = 20


#==============================================================================
#                        INITIAL
#==============================================================================

    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    cf = ax.contourf(x, y, ANOM[:, :, 2], cmap=div, levels=np.linspace(-300, 300, 13))
    #cf = ax.contourf(lon, lat, ANOM[:, :, 2], projection='mill', cmap=div, levels=np.linspace(-300, 300, 13))
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    #ax.quiver(x, y, XMEAN[:, :, 0], XMEAN[:, :, 1], width=0.001)
    q = ax.quiver(x[::2,::2], y[::2,::2], XMEAN[::2, ::2, 0], XMEAN[::2, ::2, 1], scale=60, units='inches', scale_units='inches')
    ax.quiverkey(q, 0.90, 0.1, 30, r'30 $\frac{m}{s}$', labelpos='E', coordinates='figure')
    ax.set_title('PHI (anomaly, shaded), U;V (mean, vectors)', fontsize=fstitle)
    plt.savefig(outdir + '/Initial.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                         U SPREAD
#==============================================================================
    print('    Spread')
    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ)
    cf = ax.contourf(x, y, XSPREAD[:, :, 0], cmap=seq)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('U Spread (m/s)', fontsize=fstitle)
    plt.savefig(outdir + '/U_Spread.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                        V SPREAD
#==============================================================================

    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ)
    cf = ax.contourf(x, y, XSPREAD[:, :, 1], cmap=seq)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('V Spread (m/s)', fontsize=fstitle)
    plt.savefig(outdir + '/V_Spread.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                        PHI SPREAD
#==============================================================================

    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ)
    cf = ax.contourf(x, y, XSPREAD[:, :, 2], cmap=seq)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('PHI Spread (m/s)', fontsize=fstitle)
    plt.savefig(outdir + '/PHI_Spread.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                          VAR - U COVARIANCE
#==============================================================================
    print('    Covariance')
    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(COVARIANCE[:, :, 0]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, COVARIANCE[:, :, 0], cmap=div, levels = levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('Covariance '+ VAR + '-U', fontsize=fstitle)
    plt.savefig(outdir + '/' + VAR + '-U_Covariance.png', bbox_inches='tight')

#==============================================================================
#                           VAR - V COVARIANCE
#==============================================================================

    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(COVARIANCE[:, :, 1]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, COVARIANCE[:, :, 1], cmap=div, levels = levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('Covariance '+ VAR + '-V', fontsize=fstitle)
    plt.savefig(outdir + '/' + VAR + '-V_Covariance.png', bbox_inches='tight')

#==============================================================================
#                         VAR - PHI COVARIANCE
#==============================================================================
    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(COVARIANCE[:, :, 2]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, COVARIANCE[:, :, 2], cmap=div, levels = levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('Covariance '+ VAR + '-PHI', fontsize=fstitle)
    plt.savefig(outdir + '/' + VAR + '-PHI_Covariance.png', bbox_inches='tight')

#==============================================================================
#                 VAR - PHI , VAR - U , VAR - V COVARANICE
#==============================================================================
    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(COVARIANCE[:, :, 2]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, COVARIANCE[:, :, 2], cmap=div, levels=levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    #ax.quiver(x, y, COVARIANCE[:, :, 0], COVARIANCE[:, :, 1], units='height', scale=0.6, width=0.002)
    q = ax.quiver(x, y, COVARIANCE[:, :, 0], COVARIANCE[:, :, 1], scale=wscale, units='inches', scale_units='inches')
    ax.quiverkey(q, 0.90, 0.1, wkey, wkey, labelpos='E', coordinates='figure')
    ax.set_title('Covariance '+ VAR + '-PHI (shaded), ' + VAR + '-U ; ' + VAR + '-V (vectors)', fontsize=fstitle)
    plt.savefig(outdir + '/' + VAR + '-PHI_Covariance_Vectors.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                        VAR - U LOCALIZED COVARIANCE
#==============================================================================
    print('    Localized covariance')
    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(COVARIANCE[:, :, 0]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, COVARIANCE[:, :, 0]*LOCALIZATION, cmap=div, levels=levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('Localized covariance '+ VAR + '-U', fontsize=fstitle)
    plt.savefig(outdir + '/' + VAR + '-U_Localized_Covariance.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                       VAR - V LOCALIZED COVARIANCE
#==============================================================================
    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(COVARIANCE[:, :, 1]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, COVARIANCE[:, :, 1]*LOCALIZATION, cmap=div, levels=levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('Localized covariance '+ VAR + '-V', fontsize=fstitle)
    plt.savefig(outdir + '/' + VAR + '-V_Localized_Covariance.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                     VAR - PHI LOCALIZED COVARIANCE
#==============================================================================

    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(COVARIANCE[:, :, 2]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, COVARIANCE[:, :, 2]*LOCALIZATION, cmap=div, levels=levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('Localized covariance '+ VAR + '-PHI', fontsize=fstitle)
    plt.savefig(outdir + '/' + VAR + '-PHI_Localized_Covariance.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#               VAR - PHI , VAR - U , VAR - V LOCALIZED COVARANICE
#==============================================================================

    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(COVARIANCE[:, :, 2]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, COVARIANCE[:, :, 2]*LOCALIZATION, cmap=div, levels=levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    #ax.quiver(x, y, COVARIANCE[:, :, 0]*LOCALIZATION, COVARIANCE[:, :, 1]*LOCALIZATION, units='height', scale=0.6, width=0.002)
    q = ax.quiver(x, y, COVARIANCE[:, :, 0]*LOCALIZATION, COVARIANCE[:, :, 1]*LOCALIZATION, scale=wscale, units='inches', scale_units='inches')
    ax.quiverkey(q, 0.90, 0.1, wkey, wkey, labelpos='E', coordinates='figure')
    ax.set_title('Localized covariance '+ VAR + '-PHI (shaded), ' + VAR + '-U ; ' + VAR + '-V (vectors)', fontsize=fstitle)
    plt.savefig(outdir + '/' + VAR + '-PHI_Localized_Covariance_Vectors.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                 SINGLE OBSERVATION ANALYSIS INCREMENT
#==============================================================================
    print('    Analysis increment')
    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(INCREMENT[:, :, 2]))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, INCREMENT[:, :, 2], cmap=div, levels=levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    #ax.quiver(x, y, INCREMENT[:, :, 0], INCREMENT[:, :, 1], units='height', scale=0.6, width=0.002)
    q = ax.quiver(x, y, INCREMENT[:, :, 0], INCREMENT[:, :, 1], scale=wscale, units='inches', scale_units='inches')
    ax.quiverkey(q, 0.90, 0.1, wkey, str(wkey), labelpos='E', coordinates='figure')
    ax.set_title('Analysis Increment in PHI (shaded), U;V (vectors)', fontsize=fstitle)
    plt.savefig(outdir + '/' + 'Analysis_Increment_PHI_Vectors.png', bbox_inches='tight')
    plt.close()

#==============================================================================
#                      SINGLE LOCALIZATION
#==============================================================================
    print('    Localization function')
    fig, ax, m, x, y = set_env_map(fs, lon, lat, LonI, LatJ, marker=True)
    scale = np.amax(np.abs(LOCALIZATION))/5
    levels = np.array([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5])*scale
    cf = ax.contourf(x, y, LOCALIZATION, cmap=div, levels=levels)
    ax.contour(cf, colors='k', linewidths=0.5)
    cbar = m.colorbar(cf, location='right')
    cbar.ax.tick_params(labelsize=fs)
    ax.set_title('Localization function', fontsize=fstitle)
    plt.savefig(outdir + '/' + 'Localization_function.png', bbox_inches='tight')
    plt.close()
